### ISSUE TITLE
React/Javascript translation customisations are difficult

### ISSUE THREAD

**Issue by jonom** [silverstripe/silverstripe-framework#9137]
## Affected Version

SS 4.x

## Description

Opening this issue at @robbieaverill's [suggestion](https://github.com/silverstripe/silverstripe-asset-admin/issues/765#issuecomment-497902302). From [the docs](https://docs.silverstripe.org/en/4/developer_guides/i18n/#module-priority):
> it is pretty common for a site developer to want to override the default i18n strings from time to time

In SS3 if you wanted to override a piece of text for any part of the CMS UI, you could generally just add a language yml file to your project and override the relevant key. Likewise if you wanted to translate the CMS to a new language (or add missing keys for an existing language) and see your translations in the CMS before they were accepted in Transifex and shipped in a release. Now that parts of the CMS are built in React that seems to be a lot harder ([example awful hack](https://github.com/silverstripe/silverstripe-asset-admin/issues/765#issuecomment-497511047)).

Besides being inconvenient, I think it could be confusing for developers if yml lang files work for some parts of the CMS but not others as it's inconsistent (and I don't think the limitation is documented as far as I can tell). I'm not sure if this is possible, but it would be nice if translation json files for React components were built on the fly for each project in the `flush` step, so that additional and/or custom translations were accounted for.

**Comment by robbieaverill** [comment]
Yeah we definitely need to build a JavaScript text collector

**Comment by michalkleiner** [comment]
Good call @jonom. I personally wish these things were thought of before introducing the React UI in the backend and/or before it was made the only option. So many issues with it to be resolved — this, any non-react-abled fields don't work 100%, state persistence issues, inline validation issues etc.

¯\_(ツ)_/¯

**Comment by jonom** [comment]
@michalkleiner I feel you but I think it's probably a very tough balancing act. A massive amount of work has gone in to SS4 and is still going in to it, so I think if the team had waited until everything they wanted to do was complete and super stable, we would still be waiting for SS4.0-rc1 😄 SilverStripe is going through a transition phase, like puberty. There will be some pimples and awkward moments on the way but perhaps it's a necessary step before maturity 😂

That said, I have to admit as well that I personally didn't go very deep when trying out pre-releases of SS4. If I had invested more time in trying to convert some of my existing sites to SS4 before 4.0 shipped (rather than just firing `composer create-project` and playing around) I might have been able to surface some of these issues earlier so they could be addressed before the APIs went stable. Maybe for SS5, upgrade tools could be prioritised and available before 5.0 is released. That might lead to deeper and broader (more developers and more websites) testing against real world SS4 websites and development patterns, and we might see more of these issues raised and addressed before going stable. I think part of the problem is that a large part of the community (myself included) waits until at least `.1` is released before getting seriously engaged with new major versions, as they want to wait until major issues and bugs have been ironed out. But by then we're locked in to an API and can't break it.

That was some pretty meta chatter, sorry! But maybe some points to consider in there.

**Comment by michalkleiner** [comment]
@jonom yeah, I do agree with the sentiment. But to be honest, at the same time, SS went through all that child imperfections and puberty etc. with version 3. I'd hope every major version doesn't equal a newborn baby with all its issues :-D

**Comment by maxime-rainville** [comment]
I was looking at how you would actually do this. The logic would look something like this:
* create a `lang` folder with a JS file matching your target locale (e.g.: `app/lang/en.js`)
* add the translation you want to override in your file like this
```js
    ss.i18n.addDictionary('en', {
        "AssetAdmin.ADD_FOLDER_BUTTON": "Add directory",
    });
```
* expose your `lang` folder in your `composer.json` file and run a `composer vendor-expose`
* Tell SilverStripe to look in your lang folder for translation files when requiring translations with something like this: `Requirements::add_i18n_javascript('app/lang', false);`

The tricky bit I'm not 100% sure about is where to put that `Requirements` statement, because it needs to execute after the core SilverStripe one, otherwise our custom translation file will be loaded before the official one and our translation will be over-ridden.

**Comment by jonom** [comment]
@maxime-rainville I could be wrong but if you look at the [example](https://github.com/silverstripe/silverstripe-asset-admin/issues/765#issuecomment-497511047) I posted, I think the problem is that individual components such as the UploadField re-load their own dictionary each time they are mounted, which would reset any fields that had been overridden using the technique you posted?

**Comment by robbieaverill** [comment]
That _shouldn't_ be how it works - while components will `import i18n from 'i18n'` and call `i18n._t()`, the global dictionaries are stores on `window.ss.i18n` in global state

**Comment by jonom** [comment]
> That shouldn't be how it works - while components will import i18n from 'i18n' and call i18n._t(), the global dictionaries are stores on window.ss.i18n in global state

@robbieaverill Yeah experimenting again I think I was wrong about that. I knew this was in the global state but my overrides were getting overridden and I must have made an assumption about the module dictionaries getting loaded more than once, but as @maxime-rainville suggested it was probably just an execution order issue. Have updated my [gross hack](https://github.com/silverstripe/silverstripe-asset-admin/issues/765#issuecomment-497511047) accordingly.

> The tricky bit I'm not 100% sure about is where to put that Requirements statement, because it needs to execute after the core SilverStripe one, otherwise our custom translation file will be loaded before the official one and our translation will be over-ridden.

@maxime-rainville Maybe time for Requirements methods to have an `$after` option?

**Comment by maxime-rainville** [comment]
Just a quick look and we don't provide a nice elegant way to specify the priority scripts should be loaded.

`LeftAndMain` does have a `init` hook that could be use to apply an extension to a specific controller to load additional language file. This might not work, because some of the `LeftAndMain` children load their JS in their own overloaded `init` method ... so their Requirements calls will be run after the `init` hook. See `AssetAdmin`.

`LeftAndMain` does provide a `extra_requirements_javascript` and a `extra_requirements_css` config option. I'm thinking we should introduce a `extra_requirements_i18n_javascript` option and make sure all controllers extending `LeftAndMain` use those config options rather than override the `init` method.

**Comment by maxime-rainville** [comment]
I like the "One API for translations" idea, but that would probably require a breaking change, so it would have to go in the next major release.

Adding extra parameters and/or methods to control the order things get inserted could probably be done in a minor, but that probably require a wider conversation.

Would you like to draft RFCs for those ideas.

In the short term, I think we can make it easier for people to override our translations by being a bit more aware of how we include JS files in our LeftAndMain subclasses.

**Comment by robbieaverill** [comment]
When considering the LeftAndMain ideas thrown around here, keep in mind that JavaScript translations (the same as YAML) can be used on the frontend as well as in the CMS
